Update to version 0.7.0

OnnoEbbens opened the pull request.

bdestombe commented:
Great progress!

bdestombe commented on nlmod/dims/time.py:
```diff
@@ -1,20 +1,16 @@
 
 
-def set_ds_time(
+def set_ds_time_deprecated(
```
Is it needed to keep this deprecated function?

bdestombe commented on nlmod/dims/time.py:
```diff
@@ -130,9 +133,192 @@ def set_ds_time(
+
+    if time_units == "D":
+        time_units = "DAYS"
+    ds.time.attrs["time_units"] = time_units
```
Instead import the attributes from the dims/attributes_encodings.py?
```
from dims.attributes_encodings import dims_attrs
ds.time.attrs = dims_attrs["time"]
```

And add 
```
"time":  dict(
        name="Time",
        description="Time at the end of the stress period"),
```

bdestombe replied:
See commit below

bdestombe commented on nlmod/epsg28992.py:
Could you maybe elaborate a bit? Better than which'/who's definition? And a date of the findings

OnnoEbbens replied:
I added the requested info

OnnoEbbens commented:
Thanks for the review Bas! Do you have time to look at some of the Codacy issues (the one failing test)? There are about 50 new issues and most of them can be solved easily.

dbrakenhoff commented on nlmod/dims/grid.py:
```diff
@@ -224,6 +276,47 @@ def modelgrid_to_ds(mg):
+        if mg.angrot == 0.0:
+            x += mg.xoffset  # convert to global coordinates
+            y += mg.yoffset  # convert to global coordinates
+        coords = {"layer": layers, "x": x, "y": y}
```
should we order this layer, y, x or does that not matter?

dbrakenhoff commented on nlmod/dims/layers.py:
```diff
@@ -554,24 +683,24 @@ def set_model_top(ds, top, min_thickness=0.0):
     """
     if "gridtype" not in ds.attrs:
-        raise (Exception("Make sure the Dataset is build by nlmod"))
+        raise (KeyError("Make sure the Dataset is built by nlmod"))
```
Perhaps add a suggestion which method to use to build the dataset here?

dbrakenhoff commented on nlmod/dims/layers.py:
```diff
@@ -807,77 +963,134 @@ def fill_nan_top_botm_kh_kv(
 
-def fill_top_and_bottom(ds):
-    """Remove Nans in botm variable, and change top from 3d to 2d if needed."""
+def fill_top_and_bottom(ds, drop_layer_dim_from_top=True):
```
This name is a bit vague considering the change from 3D to 2D top is also done. Maybe we can call this ` to_modflow_tops_and_bottoms()` which calls 2 functions, the first being `fillnan_top_and_bottom()` and the second `make_top_2d()`?

OnnoEbbens replied:
I agree in splitting this function. The code to fill nan values in the top and bottom can be optimized using the xarray bfill or ffill methods. I think we should create a separate issue for this

dbrakenhoff replied:
I created an issue for this, so we can resolve this conversation.

dbrakenhoff commented on nlmod/dims/layers.py:
```diff
@@ -1040,3 +1321,271 @@ def aggregate_by_weighted_mean_to_ds(ds, source_ds, var_name):
+    method `nlmod.read.get_regis()`, and does not function for a model Dataset with a 2d
+    (structured) or 1d (vertex) top.
+
+    When comparing the height of the new layer with an existing layer, there are 7
```
there are 6 items below?

dbrakenhoff commented on nlmod/dims/layers.py:
```diff
@@ -1040,3 +1321,271 @@ def aggregate_by_weighted_mean_to_ds(ds, source_ds, var_name):
+    6 The new layer is below the existing layer everywhere. Nothing happens, move on to
+    the next existing layer.
+
+    When (part of) the new layer is not added to the layer model after comparison
```
Ah, this should be 7

dbrakenhoff commented on nlmod/gwf/gwf.py:
```diff
@@ -499,8 +485,176 @@ def drn(
+    layer=0,
+    **kwargs,
+):
+    """create constant head boundary at the model's edges from the model dataset.
```
This description is wrong I think it's no longer only model edges, but wherever the mask says it should be.

dbrakenhoff commented on nlmod/gwf/recharge.py:
```diff
@@ -154,15 +200,318 @@ def ds_to_evt(gwf, ds, pname="evt", nseg=1, surface=None, depth=None, **kwargs):
+        when it is None. The default is None.
+    finf :float, str or array-like
+        The applied infiltration rate of the UZF cell (:math:`LT^{-1}`). When passed as
+        string, the array is obtained from ds. The default is "recharge".
```
Do we want to name this "recharge"  by default or should it be precipitation? Recharge implies we already include evaporation somehow?

OnnoEbbens replied:
We can also set the default to None so the user has to decide what data should be used?

dbrakenhoff commented on nlmod/util.py:
```diff
@@ -15,10 +15,50 @@
+    pass
+
+
+class MissingValueError(Exception):
```
Maybe add a description where we use these errors? Or move them to the appropriate file?

OnnoEbbens replied:
I agree

dbrakenhoff commented on nlmod/util.py:
```diff
@@ -411,14 +455,24 @@ def download_mfbinaries(bindir=None):
 
 def download_modpath_provisional_exe(bindir=None, timeout=120):
-    """Downlaod the provisional version of modpath to the folder with binaries"""
+    """Downlaod the provisional version of modpath to the folder with binaries."""
```
typo

dbrakenhoff reviewed (approved):
Lot's of great improvements :)! I made a few minor comments, nothing major. These could also be converted into issues to deal with in a next version.

dbrakenhoff commented:
I think we should just use our admin powers to ignore issues on codacy, or somehow add some of the issues it raises to the ignore list. Most of them aren't really an issue or worth fixing.

dbrakenhoff commented:
Has #261 been fixed yet? It would be good to get that one done before this release I think.

OnnoEbbens commented:
I fixed most of the codacy errors. Some ideas/remarks/questions:
- I disabled the `too_complex` pattern and the `too_many_branches` patterns because that are often hard to solve. However I think the checks are good indicators of code that is hard to read so I feel conflicted about disabling them.
- While looking through some files I notice many functions miss a docstring. See functions in `dcs.py`, `geotop.py` and `regis.py`. I think we should enforce docstrings.
- @bdestombe: `{}` is preferred over `dict()` according to the codacy style checker. I already fixed this but I think it is good to know for future commits.
- @dbrakenhoff: can you check this codacy error, maybe replace `pass` with `...`:
```
nlmod/util.py
Medium icon Medium Code Style
Unnecessary pass statement

pass
```
- @dbrakenhoff: can you check this codacy error, I cannot figure out if we can ignore this:
```
nlmod/dims/base.py
Minor icon Minor Code Style
bad operand type for unary ~: NoneType (invalid-unary-operand-type)

data[lay, mask] = data[lay, ~mask][i]
```
- @dbrakenhoff: probably the same as the previous one:
```
nlmod/dims/base.py
Minor icon Minor Code Style
bad operand type for unary ~: NoneType (invalid-unary-operand-type)

data[mask] = data[~mask, i]
```
- @rubencalje: can you check this codacy error (and add a docstring for the `dataframe_to_flopy_timeseries` function?):
 ```
nlmod/dims/time.py
Minor icon Minor Code Style
Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)

def dataframe_to_flopy_timeseries(
```
- @rubencalje: can you check this codacy error? Should we add skimage as a dependency?
```
nlmod/read/rws.py
Minor icon Minor Code Style
Import outside toplevel (skimage.morphology.reconstruction) (import-outside-toplevel)

from skimage.morphology import reconstruction
```

OnnoEbbens commented:
Suddenly a new check: `too-many_statements` appeared. I decided to ignore that one

dbrakenhoff commented:
The ~ issues can be ignored. I'm not sure how to avoid the check being triggered for these cases.
